Gradle: Align the Gradle version during Arquillian dependency resolution

roguexz opened the pull request:
Previously, the Gradle Arquillian integration would always spin up a Gradle 3.1 daemon. This was becuase we were pulling in the Gradle jar (version 3.1) as a dependency. This has a performance implication on the CI equation as each build would download Gradle 3.1 for testing.

With this commit, I am making the following changes,

1. Mark all Gradle dependencies as provided.
2. Include the required dependencies (from the Gradle platform) dynamically as part of the plugin configurations.
3. Introduce a path forward for adding Gradle specific changes without having to rely on bumping the Gradle library versions.

-----

- [x] Have you followed the guidelines in our [Contributing](https://thorntail.io/community/contributing/) document?
- [ ] Have you created a [JIRA](https://issues.jboss.org/browse/THORN) and used it in the commit message?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/thorntail/thorntail/pulls) for the same issue?
- [x] Have you built the project locally prior to submission with `mvn clean install`?

-----

roguexz committed:
Gradle: Align the Gradle version during Arquillian dependency resolution

Previously, the Gradle Arquillian integration would always spin up a Gradle 3.1 daemon. This was becuase we were pulling in the Gradle jar (version 3.1) as a dependency. This has a performance implication on the CI equation as each build would download Gradle 3.1 for testing.

With this commit, I am making the following changes,

1. Mark all Gradle dependencies as provided.
2. Include the required dependencies (from the Gradle platform) dynamically as part of the plugin configurations.
3. Introduce a path forward for adding Gradle specific changes without having to rely on bumping the Gradle library versions.

roguexz commented:
@Ladicek @TomasHofman 
Another one for you to review.
I have started looking at performance improvements for the Gradle plugin. As I find out more, I will keep contributing them back.

FYI, I have the changes done for the examples project as well. Just want to make one more change that has a huge impact on the performance of testing the plugins. If I unable to make much progress on that front, I will raise the PR for examples project anyways.

Ladicek commented on plugins/gradle/gradle-plugin/src/main/java/org/wildfly/swarm/plugin/gradle/ThorntailArquillianPlugin.java:
```diff
@@ -44,11 +58,65 @@ public ThorntailArquillianPlugin(ToolingModelBuilderRegistry registry) {
+                        File scriptFile = new File(project.getBuildDir(), "thorntail-arquillian-script.gradle");
+                        if (!scriptFile.exists()) {
+                            scriptFile.createNewFile();
+                            String[] lines = readString(resource.openStream()).split("\n");
+                            Files.write(scriptFile.toPath(), Arrays.asList(lines));
```
Why split the string into lines if you immediately merge them back?

roguexz replied:
So, when I passed in a string directly (which had `\n`), it messed up the content in the file. I tried seeing if there was a convenience method in the codebase, but was unable to locate one. If you know of something, then I can switch to that.

Ladicek commented on plugins/gradle/gradle-plugin/src/main/java/org/wildfly/swarm/plugin/gradle/ThorntailArquillianPlugin.java:
```diff
@@ -44,11 +58,65 @@ public ThorntailArquillianPlugin(ToolingModelBuilderRegistry registry) {
+     * @param stream the input stream.
+     * @return the string built out of the data available in the given stream.
+     */
+    private static String readString(InputStream stream) throws IOException {
```
Isn't it sad that there's no equivalent of `Files.readAllBytes` for classpath resources? :-/

roguexz replied:
I know! Additionally, the `Files` API, works only with files and not resources inside a Jar or other locations.

Ladicek commented on plugins/gradle/gradle-plugin/src/main/resources/META-INF/thorntail-scripts/configure-tasks.groovy:
```diff
@@ -0,0 +1,29 @@
+ */
+
+// Given our current integration model, we end up in a scenario where we are unable to perform all integrations from
+// Java layer as that would mean that we now need to have a dependency on a newer version of the Gradle plugin.
```
Hmm... I briefly worked with Gradle around their 1.0 time, and I think that even back then, their Java API already exposed enough to find all tasks of type `Test` in the project. In fact, I think the Groovy scripts use the same API that you can use from Java. So I'm not quite sure that we need to add an extra Groovy script; but of course you have much more experience with Gradle than me :-)

roguexz replied:
So the problem that I faced was the following,

```
Test (task) --(references)-- DefaultJavaForkOptions
DefaultJavaForkOptions --(extends)--> DefaultProcessForkOptions --(implements)--> ProcessForkOptions
```

Now, `ProcessForkOptions` isn't available on the classpath, which causes the compilation to fail (as it is an internal API). Since Gradle doesn't ship an API compatible jar that we can use, we end up having to resort to a runtime resolution instead of relying at Reflection at compile time. Between the two choices, the runtime option keeps things simple. I was planning on removing all the reflection code in favor of clear concise Groovy code. Hence the decision to go with this approach. But I am definitely open to suggestions.

Ladicek replied:
Understood, thanks!

Ladicek commented on plugins/gradle/tooling-api/src/main/java/org/wildfly/swarm/plugin/gradle/GradleToolingHelper.java:
```diff
@@ -27,6 +27,17 @@
+     * of the project against a newer version of Gradle instead of having to upgrade the Gradle wrapper.properties file.
+     * <p>
+     * Also, useful when the project root folder does not contain the Gradle wrapper properties, as is the case with the
+     * Thorntail Gradle examples. Since we cannot anticipate all possible scenarios, it is advisable to use the same
+     * version of Gradle that started this build.
```
+1

Ladicek reviewed (dismissed):
Looks good. Just added a few minor comments, but nothing blocking.

Ladicek commented:
I'll wait a bit to give you a chance to reply to my comments, but I'd be fine with merging as is.

roguexz committed:
Create the build directory if it doesn exist.

The Thorntail Arquillian plugin tries to create directories before any of the other project plugins have kicked in (e.g., compile, assemble, etc.). In such cases, we need to ensure that the build directory exists before attempting to create a file in it.

roguexz commented on plugins/gradle/gradle-plugin/src/main/java/org/wildfly/swarm/plugin/gradle/ThorntailArquillianPlugin.java:
```diff
@@ -44,11 +58,68 @@ public ThorntailArquillianPlugin(ToolingModelBuilderRegistry registry) {
+                URL resource = ThorntailArquillianPlugin.class.getClassLoader().getResource(BUILD_SCRIPT_FRAGMENT);
+                if (resource != null) {
+                    try {
+                        if (!project.getBuildDir().exists()) {
```
Resolved a small bug that would have become a show stopper :) .. All thanks to the Thorntail Examples. I should probably add a note to the README reminding myself and others that we should verify the plugin changes against the example projects before raising a PR.

roguexz committed:
Apply the thorntail-arquillian plugin properly

This commit fixes a bug where the `thorntail` Gradle plugin wasn't applying the `thorntail-arquillian` plugin properly.

roguexz commented:
@Ladicek I am done with the changes for this PR. I have verified the changes in the example projects. Additional PRs will only reduce the amount of code that we need to write in the build scripts.
